src/cipherstate.rs

```diff
@@ -192,3 +184,46 @@ impl StatelessCipherStates {
         self.1.rekey_manually(key)
     }
 }
+
+trait HasKey {
```

Comment:
I see what you're doing with this trait, removing duplicate code in CipherState and StatelessCipherState, and I think it's a fitting change for another PR. But I think I'd like to keep this PR simple for now and have some duplicate code between the structs (of which there is already a good amount):

```rust
        if (ciphertext.len() < TAGLEN) || buffer.len() < (ciphertext.len() - TAGLEN) {
            return Err(Error::Decrypt);
        }

        if !self.has_key() {
            return Err(StateProblem::MissingKeyMaterial.into());
        }
```

Reply:
I reverted the trait cleanup on the branch. I'll look at a different PR for that after this one :)